File: DoublyLinkedList/DLL.cpp
#include<iostream>
#include <cstdlib>
using namespace std;

class Node{
    public:
      int data;
      Node *next;
      Node *prev;
      Node(int data){
          this->data=data;
          this->next=NULL;
          this->prev=NULL;
      } 
};

class DLL{
    public:
      Node *head;
      DLL(){
          this->head=NULL;
      }
      void addAtFront(int data){
          Node *newNode=new Node(data);
          if(this->head==NULL){
              this->head=newNode;
              return;
          }
          newNode->next=this->head;
          this->head->prev=newNode;
          this->head=newNode;
          return;
      }

      void addAtEnd(int data){
          Node *newNode=new Node(data);
          if(this->head==NULL){
              this->head=newNode;
          }
          Node *current=this->head;
          while(current->next!=NULL){
              current=current->next;
          }
          newNode->prev=current;
          current->next=newNode;
      }
      
      void deleteFront(){
         if(this->head==NULL){
             cout<<"No Node to delete\n";
             return;
         } 
         Node *del=this->head;
         del->next->prev=NULL;
         this->head=del->next;

         free(del);
      }

      void deleteEnd(){
          if(this->head==NULL){
            cout<<"No Node to delete\n";
            return;
          }

          Node *current=this->head;
          while(current->next!=NULL){
              current=current->next;
          }
          current->prev->next=NULL;
          free(current);
      }

      void reverse(){
          Node *current=this->head;
          Node *temp=NULL;
          while(current!=NULL){
              temp=current->next;
              current->next=current->prev;
              current->prev=temp;
              temp=current;
          }
          this->head=temp;
      }

      void printAllBack(){
        Node *current=this->head;
        while(current->next!=NULL){
            current=current->next;
        }

        while(current!=NULL){
            cout<<current->data<<" ";
            current=current->prev;
        }
      }

      void printAll(){
         Node *current=this->head;
         while(current!=NULL){
             cout<<current->data<<" ";
             current=current->next;
         }
      }

};

int main(){
         DLL *head=new DLL();
         
         head->addAtFront(2);
         head->addAtFront(5);
         head->addAtFront(10);
         head->addAtEnd(20);

         cout<<"\nPrint From Start to End\n";
         head->printAll();

        //  head->deleteFront();

        //  head->deleteEnd();
         head->reverse();
         cout<<"\nPrint From Start to End\n";
         
         head->printAll(); 
        //  cout<<"\nPrint From End To Start\n"; 
        //  head->printAllBack();

         return 0;
         }

File: DoublyLinkedList/DLL.js
class Node{
    constructor(data){
        this.data=data
        this.next=null
        this.prev=null
    }
}

class DLL{
    constructor(){
        this.head=null
    }

    addToStart(data){
        let newNode=new Node(data)
        if(this.head==null){
            this.head=newNode
            return
        }

        newNode.next=this.head
        this.head.prev=newNode
        this.head=newNode
    }

    addAtEnd(data){
       let newNode=new Node(data)
       if(this.head==null){
           this.head=newNode
           return
       }

       let current=this.head
       while(current.next!=null){
           current=current.next
       }
       newNode.prev=current
       current.next=newNode
    }
    deleteFromStart(){
        if(this.head==null){
            console.log("List Is Empty")
            return
        }
        this.head.next.prev=null
        this.head=this.head.next
    }

    deleteFromEnd(){
        let current=this.head
        while(current.next.next!=null){
            current=current.next
        }
        current.next=null
    }
    printAll(){
        let current=this.head
        while(current!=null){
            console.log(current.data)
            current=current.next
        }
    }

    printReverse(){
        let current=this.head
        while(current.next!=null){
            current=current.next
        }
        while(current!=null){
            console.log(current.data)
            current=current.prev
        }
    }
}

let head=new DLL()

head.addToStart(10)
head.addToStart(8)
head.addToStart(6)
head.addToStart(4)
head.addToStart(2)

head.addAtEnd(12)

head.deleteFromStart()

head.deleteFromEnd()

console.log("Print in Correct Order")
head.printAll()

console.log("Print in Reverse Order")
head.printReverse()